### PR TITLE
Domains: make unverified emails more visible

### PR DESCRIPTION
**Pull request by dzver**
1. Add red warnings to domains/manage/site for unverified emails
2. Add verification component to domain details

We are required to suspend domains with unverified emails. For some tlds this is enforced outside of our control. This makes email verification equally important to domain expiration - it can break the domain if not resolved promptly.

screenshot 2016-01-29 21 19 04
screenshot 2016-01-29 21 21 52
screenshot 2016-01-29 21 23 07

To test you'll need to own several domains and set their emails to something random. You can also make one of the domains expiring or expired to see one of the other notices.

CC: @breezyskies @ranh @umurkontaci

### CODE REVIEW

File: client/my-sites/upgrades/components/domain-warnings/index.jsx [hunk 1]
```diff
@@ -93,7 +94,32 @@ export default React.createClass( {
...
+				this.translate( 'Some of your domains have unverified email addresses. This may lead to their temporary suspension.' )
+			}</Notice>;
+		} else if ( domains.length > 0 ) {
+			notices = domains.map( domain => {
+				const text = this.translate( 'You need to verify the email for %(domainName)s or the domain may be suspended. {{a}}Learn more.{{/a}}', {
+					args: {
+						domainName: domain.name
+					},
+					components: {
+						a: <a href={ paths.domainManagementEdit( this.props.selectedSite.domain, domain.name ) } />
```
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: The link says learn more but it goes to the edit page. Maybe we should change it naming or the href itself?

File: client/my-sites/upgrades/components/domain-warnings/index.jsx [hunk 2]
```diff
@@ -93,7 +94,32 @@ export default React.createClass( {
...
+					}
+				} );
+				return <Notice status="is-error" showDismiss={ false } key={ 'unverifiedDomains' + domain.name }>{ text }</Notice>;
+			} );
+		}
+
+		return notices ? <div key="unverifiedDomains">{ notices }</div> : null;
```
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: No need for ternary here: `return !! notices || <div key="unverifiedDomains">{ notices }</div>;`

File: client/my-sites/upgrades/components/domain-warnings/index.jsx [hunk 3]
```diff
@@ -179,15 +205,8 @@ export default React.createClass( {
...
-		let pipe = this.getPipe();
-
-		for ( let renderer of pipe ) {
-			let result = renderer();
-			if ( result ) {
-				return result;
-			}
-		}
-		return null;
+		const notices = this.getPipe().map( renderer => renderer() ).filter( notice => notice );
```
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: You're changing the behaviour to show all the warnings at once. Why?

File: client/my-sites/upgrades/components/domain-warnings/index.jsx [hunk 4]
```diff
@@ -179,15 +205,8 @@ export default React.createClass( {
...
+		return notices.length ? <div>{ notices }</div> : null;
```
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: Same thing, no need for a ternary

File: client/my-sites/upgrades/domain-management/edit/registered-domain.jsx [hunk 5]
```diff
@@ -192,6 +193,8 @@ const RegisteredDomain = React.createClass( {
 						<SubscriptionSettings
 							onClick={ this.handlePaymentSettingsClick } />
 					</Card>
+
+					{ domain.isPendingIcannVerification ? <IcannVerificationCard selectedDomainName={ domain.name } /> : null }
```
Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/edit/registered-domain.jsx]: `{ domain.isPendingIcannVerification && <IcannVerificationCard selectedDomainName={ domain.name } /> }`

File: client/my-sites/upgrades/domain-management/name-servers/index.jsx [hunk 6]
```diff
@@ -11,7 +11,7 @@ import Main from 'components/main';
 import Header from 'my-sites/upgrades/domain-management/components/header';
 import CustomNameserversForm from './custom-nameservers-form';
 import WpcomNameserversToggle from './wpcom-nameservers-toggle';
-import IcannVerificationCard from './icann-verification-card';
+import IcannVerificationCard from '../components/icann-verification-card';
```
Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/name-servers/index.jsx]: You should use absolute paths if the file you're importing is not in the same folder.

File: client/my-sites/upgrades/domain-management/style.scss [hunk 7]
```diff
@@ -848,20 +848,18 @@ ul.email-forwarding__list {
 	}
 }
 
-.name-servers {
-	.icann-verification {
-		.button {
-			margin-bottom: 15px;
+.icann-verification {
```
Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/style.scss]: I think verification in the global css namespace is likely to have name clash in the future. We should namespace it.

File: client/my-sites/upgrades/components/domain-warnings/index.jsx [hunk 8]
```diff
@@ -93,7 +94,32 @@ export default React.createClass( {
 			} );
 			renewLink = renewLinkPlural;
 		}
-		return <Notice status="is-error" showDismiss={false}>{text} {renewLink}</Notice>;
+		return <Notice status="is-error" showDismiss={ false } key="expiringDomains">{ text } { renewLink }</Notice>;
+	},
+
+	unverifiedDomains() {
+		var notices,
```
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: Use `let`, `const` over `var`.

**Comment by umurkontaci** [comment]
This should fix #2526

**Comment by folletto** [comment]
Before / After screenshot? :)

**Comment by umurkontaci** [comment]
Can you add more context on what's going on here? E.g. why are we doing this change, screenshots and the testing instructions?

**Comment by umurkontaci** [comment]
- Shouldn't these be merged into one?
image

- I only see them on the Domains list page, I'm not sure if that's the intended behaviour though.
- @breezyskies for the warning copies

File: client/my-sites/upgrades/components/domain-warnings/index.jsx (hunk 3 above)
Comment (dzver) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: Because you can have 2 or 3 or 5 critical messages. If it's too ugly to display 5 critical notices one below the other, we should improve this component instead of hiding 4 of them.

File: client/my-sites/upgrades/domain-management/style.scss (hunk 7 above)
Comment (dzver) [inline client/my-sites/upgrades/domain-management/style.scss]: Why?

Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/style.scss]: "is likely to have name clash in the future"

File: client/my-sites/upgrades/components/domain-warnings/index.jsx (hunk 3 above)
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: It was a deliberate decision to show the most important message. If an not-validated email is more important than expiring, you can move that up.

While I agree that it will be good addition to allow multiple message to be visible at any time, it will break the current behaviour and potentially increase the complexity to use.

Comment (dzver) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: I'm proposing this for real. If 1 domain is expiring tomorrow, another expired a week ago, and 2 more are unverified and will be suspended in 5 days, all four of the messages are critical.

File: client/my-sites/upgrades/domain-management/style.scss (hunk 7 above)
Comment (dzver) [inline client/my-sites/upgrades/domain-management/style.scss]: I move IcannVerification as a separate component in this PR, it is of equal importance with .name-servers. I can't imagine why would anybody implement another similar component called IcannVerification.

Comment (umurkontaci) [inline client/my-sites/upgrades/domain-management/style.scss]: I wouldn't support `.name-servers` either. This means that this css class will be applied whenever I use `.icann-verification`, it doesn't have to be a component, just any DOM element with that class, wherever it is.

File: client/my-sites/upgrades/components/domain-warnings/index.jsx (hunk 3 above)
Comment (umurkontaci) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: If you are proposing a breaking change, I'd suggest to put this in a new PR get some designer review first. This is out of scope for this PR.

Comment (dzver) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: What do you mean by a breaking change? That after this PR the warning will not work?

**Comment by apeatling** [comment]
Can we merge the duplicate notices?

**Comment by breezyskies** [comment]
Proposal for merging multiple notices / actions:

screen shot 2016-01-29 at 12 53 46 pm

This could potentially get long if the user has a lot of domains, but that should be pretty edge case.

cc: @ranh for copy input

**Comment by dzver** [comment]
The fix link leads to domain details page? The IcannVerification component can't get any shorter, people need to know that it's 50/50 between having the verification email in spam or having a typo in their email.

File: client/my-sites/upgrades/components/domain-warnings/index.jsx [hunk 9]
```diff
@@ -93,7 +94,32 @@ export default React.createClass( {
...
+			domains = this.getDomains().filter( domain => domain.isPendingIcannVerification );
+
+		if ( domains.length > 2 ) {
+			notices = <Notice status="is-error" showDismiss={ false } key="unverifiedDomains">{
+				this.translate( 'Some of your domains have unverified email addresses. This may lead to their temporary suspension.' )
```
Comment (stephanethomas) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: Any reason why the `Learn more` link isn't included in this case?

File: client/my-sites/upgrades/components/domain-warnings/index.jsx (hunk 2 above)
Comment (stephanethomas) [inline client/my-sites/upgrades/components/domain-warnings/index.jsx]: You could also do the following - which is shorter:

```javascript
return notices && <div key="unverifiedDomains">{ notices }</div>;
```

File: client/my-sites/upgrades/domain-management/components/icann-verification-card.jsx [hunk 10]
```diff
@@ -15,15 +15,25 @@ let IcannVerificationCard = React.createClass( {
...
+	getExplanation() {
+		switch ( this.props.explanationContext ) {
+			case 'name-servers':
+				return this.translate(
+						'You must verify your email address through the ICANN ' +
+						'verification email before you are able to update the name ' +
+						'servers for your domain.' );
+		}
+
+		return this.translate(
```
Comment (stephanethomas) [inline client/my-sites/upgrades/domain-management/components/icann-verification-card.jsx]: What is the reason for not having this `return` statement in the `switch` under `default:`?

**Comment by ranh** [comment]
Suggesting some more "urgent" copy:

One domain:

> Urgent! Your domain example.com may be lost forever because your email address is not verified. [Fix now](#)

While we can't "fix" from the notice itself as Veselin mentioned about, I think that "fix now" is still an appropriate label for the link since that is the action the user should have in mind. The flow would be: 

1. *Notice* the notice (haha)
2. Hit "fix now"
3. Review instructions to fix
4. Find the email/resend the email
5. Hit link to verify

Multiple domains:

> Urgent! Some of your domains may be lost forever because your email address is not verified:
>
> * example.com [Fix now](#)
> * example.org [Fix now](#)

---

The copy on the edit domain screen could also use some touching up. Currently:

> You must verify your email address through the ICANN verification email or your domain may be suspended
> 
> [Resend ICANN verification email](#)
>
> You might be seeing this message, because the email in your contact information is wrong. 

Suggested:

> Urgent! Please verify your address through the verification email as soon as possible, or you may lose this domain forever. 
>
> [Resend verification email](#)
> 
> Use this button to resend the verification email. It contains a link to verify your address. To update your email address, go to Contacts and Privacy . 

It might also be a good idea to show a screenshot of the email here.

@breezyskies @dzver

**Comment by breezyskies** [comment]
+1 for more urgent copy and removing the references to ICANN.

> Urgent! Please verify your address through the verification email as soon as possible, or you may lose this domain forever.

> Resend verification email

> Use this button to resend the verification email. It contains a link to verify your address. To update your email address, go to Contacts and Privacy .

I might shorten this one up to one sentence + relevant action ("resend email"). If it takes more than that, we should probably link out to help docs, as these notices are not really designed for much more than a sentence or so. Maybe we could show the email address on file right in the message to make it more clear if it's wrong? Something like: 

> Resend Email to myemail@probablyhasatypo.com | Update Email

**Comment by ranh** [comment]
@breezyskies LGTM, the whole "Use this button" paragraph can go. Adding the email address is a nice touch.

**Comment by rickybanister** [comment]
Did this get handled in another PR or is something else holding it up?

**Comment by dzver** [comment]
I decided to split the PR in two parts and already merged half of it. The other half is waiting. Priorities dragged me in other directions. I hope I'll have time for it soon.

**Comment by dzver** [comment]
I created another PR with the second half of this change. It respects the proposed design and copy. Closing this.
